From 7acdb22700571c86807160d855fee43afeb84926 Mon Sep 17 00:00:00 2001
From: Yann Gautier <yann.gautier@foss.st.com>
Date: Mon, 18 Dec 2023 14:22:56 +0100
Subject: [PATCH 441/841] feat(stm32mp2): stub PM code in serial boot

When booting from USB or UART, with programmer, the low-power sequences
won't be used, some code can then be stubbed under UART and USB flags.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/arm-trusted-firmware.git 8c0e82e714694a2ad8495a4e6f4586d082a804dc feat(stm32mp2): stub PM code in serial boot]
Change-Id: I10bda7930bd809640f2b40fe46a6fa568946c09d
Signed-off-by: Yann Gautier <yann.gautier@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/tf-a/+/347470
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Reviewed-by: Sebastien PASDELOUP <sebastien.pasdeloup-ext@st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
---
 plat/st/stm32mp2/bl2_plat_setup.c   | 12 +++++++++++-
 plat/st/stm32mp2/stm32mp2_private.c |  4 ++++
 2 files changed, 15 insertions(+), 1 deletion(-)

diff --git a/plat/st/stm32mp2/bl2_plat_setup.c b/plat/st/stm32mp2/bl2_plat_setup.c
index 9ee4d855f..b061e39e8 100644
--- a/plat/st/stm32mp2/bl2_plat_setup.c
+++ b/plat/st/stm32mp2/bl2_plat_setup.c
@@ -401,7 +401,7 @@ skip_console_init:
 	}
 #endif
 
-#if STM32MP_DDR_FIP_IO_STORAGE
+#if STM32MP_DDR_FIP_IO_STORAGE && !STM32MP_UART_PROGRAMMER && !STM32MP_USB_PROGRAMMER
 	/* Skip DDR FW ID = the first image to load for standby exit */
 	if (stm32mp_is_wakeup_from_standby()) {
 		bl_mem_params_node_t *bl_mem_params;
@@ -420,6 +420,11 @@ static void prepare_encryption(void)
 #if !STM32MP_M33_TDCID
 	uint8_t mkey[RISAF_KEY_SIZE_IN_BYTES];
 
+#if STM32MP_UART_PROGRAMMER || STM32MP_USB_PROGRAMMER
+	if (stm32_rng_read(mkey, RISAF_KEY_SIZE_IN_BYTES) != 0) {
+		panic();
+	}
+#else /* STM32MP_UART_PROGRAMMER || STM32MP_USB_PROGRAMMER */
 	if (stm32mp_is_wakeup_from_standby()) {
 		stm32mp_pm_get_enc_mkey_from_context(mkey);
 	} else {
@@ -430,6 +435,7 @@ static void prepare_encryption(void)
 
 		stm32mp_pm_save_enc_mkey_in_context(mkey);
 	}
+#endif /* STM32MP_UART_PROGRAMMER || STM32MP_USB_PROGRAMMER */
 
 	if (stm32mp2_risaf_write_encryption_key(RISAF4_INST, mkey) != 0) {
 		panic();
@@ -475,7 +481,11 @@ int bl2_plat_handle_post_image_load(unsigned int image_id)
 				FW_CONFIG_ID);
 		fconf_populate("FW_CONFIG", STM32MP_FW_CONFIG_BASE);
 
+#if STM32MP_UART_PROGRAMMER || STM32MP_USB_PROGRAMMER
+		wakeup_standby = false;
+#else /* STM32MP_UART_PROGRAMMER || STM32MP_USB_PROGRAMMER */
 		wakeup_standby = stm32mp_is_wakeup_from_standby();
+#endif /* STM32MP_UART_PROGRAMMER || STM32MP_USB_PROGRAMMER */
 
 		/* Iterate through all the fw config IDs */
 		for (i = 0U; i < ARRAY_SIZE(image_ids); i++) {
diff --git a/plat/st/stm32mp2/stm32mp2_private.c b/plat/st/stm32mp2/stm32mp2_private.c
index 88a1e5396..bac139e97 100644
--- a/plat/st/stm32mp2/stm32mp2_private.c
+++ b/plat/st/stm32mp2/stm32mp2_private.c
@@ -505,6 +505,9 @@ bool stm32mp_skip_boot_device_after_standby(void)
 
 bool stm32mp_is_wakeup_from_standby(void)
 {
+#if STM32MP_UART_PROGRAMMER || STM32MP_USB_PROGRAMMER
+	return false;
+#else /* STM32MP_UART_PROGRAMMER || STM32MP_USB_PROGRAMMER */
 	static int standby = -1;
 	uint32_t rstsr;
 
@@ -522,6 +525,7 @@ bool stm32mp_is_wakeup_from_standby(void)
 	}
 
 	return standby == 1;
+#endif /* STM32MP_UART_PROGRAMMER || STM32MP_USB_PROGRAMMER */
 }
 
 int stm32_risaf_get_instance(uintptr_t base)
-- 
2.47.2

